### PR TITLE
[Feature] Define TypeInformation based on TypingResult

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/typeinformation/internal/typedobject/TypedObjectBasedTypeInformation.scala [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+    }
+  }
+
+  override def readSnapshot(readVersion: Int, in: DataInputView, userCodeClassLoader: ClassLoader): Unit = {
+    val size = in.readInt()
+    serializersSnapshots = (0 until size).map { _ =>
+      val key = in.readUTF()
+      val snapshot = InstantiationUtil.resolveClassByName(in, userCodeClassLoader)
+        .getConstructor().newInstance().asInstanceOf[TypeSerializerSnapshot[_]]
+      snapshot.readSnapshot(1, in, userCodeClassLoader)
```
Comment (arkadius) [inline engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/typeinformation/internal/typedobject/TypedObjectBasedTypeInformation.scala:143]: Is this `1` is correct? Shouldn't we use `TypeSerializerSnapshot.readVersionedSnapshot` instead? Many serializers currently uses other versions.

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/typeinformation/internal/typedobject/TypedObjectBasedTypeInformation.scala [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+      (key, snapshot)
+    }.toList
+  }
+
+  /*
+    The idea is as follows:
+    if nonEqualKeysCompatible == true we check if keys present both in new and old serializer are compatible. If
```
Comment (arkadius) [inline engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/typeinformation/internal/typedobject/TypedObjectBasedTypeInformation.scala:150]: +1 for this flag and for description, but I don't like negation in its name. WDYT about negating it and name like `compatibleWhenAllKeysSame` or similar?

File: engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/typeinformation/TypingResultAwareTypeInformationDetectionSpec.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+import org.apache.flink.api.common.ExecutionConfig
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.api.common.typeutils.TypeSerializerSchemaCompatibility
+import org.apache.flink.core.memory.{DataInputViewStreamWrapper, DataOutputViewStreamWrapper}
+import org.scalatest.{FunSuite, Matchers}
+import pl.touk.nussknacker.engine.api.{Context, ValueWithContext}
+import pl.touk.nussknacker.engine.api.context.ValidationContext
+import pl.touk.nussknacker.engine.api.typed.typing.{Typed, TypedObjectTypingResult, TypingResult}
+
+class TypingResultAwareTypeInformationDetectionSpec extends FunSuite with Matchers {
```
Comment (arkadius) [inline engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/typeinformation/TypingResultAwareTypeInformationDetectionSpec.scala:14]: I wonder if you can add some tests showing problems that can occur during using this dection. E.g. you mention about Int vs Long. What about tests for `nonEqualKeysCompatible` feature?

**Review by arkadius — COMMENTED** [review]
Basically LGTM, I've added some minor comments that I would like to discuss.

**Review by arkadius — APPROVED** [review]

File: engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/typeinformation/internal/typedobject/TypedObjectBasedTypeInformation.scala [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        target.writeBoolean(true)
+      } else {
+        target.writeBoolean(false)
+        serializer.asInstanceOf[TypeSerializer[AnyRef]].serialize(valueToSerialize, target)
+      }
+    }
+  }
+
+  override def deserialize(source: DataInputView): T = {
+    //TODO: remove array allocation
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/flink/process/src/main/scala/pl/touk/nussknacker/engine/process/typeinformation/internal/typedobject/TypedObjectBasedTypeInformation.scala]: I added comment

File: engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/typeinformation/TypingResultAwareTypeInformationDetectionSpec.scala (hunk 3 above)
Comment (mproch) [reply engine/flink/process/src/test/scala/pl/touk/nussknacker/engine/process/typeinformation/TypingResultAwareTypeInformationDetectionSpec.scala:19]: I added additional tests

**Pull request merged**